### ISSUE TITLE
cycle detected when using custom vcpkg registration with shared dependencies

### ISSUE THREAD

**Issue by niyue** [microsoft/vcpkg#20503]
**Describe the bug**
* vcpkg reported "Cycle detected" if some dependency is shared by multiple packages in custom registry.

**Environment**
- OS: macOS Big Sur (11.6)
- Compiler: Apple Clang 13.0

**To Reproduce**
Attached is the sample project `cycle-detected` used for reproducing this issue:
[cycle_detected.zip](https://github.com/microsoft/vcpkg/files/7276967/cycle_detected.zip)

1. Create a custom vcpkg registry on file system
2. Add two packages to the custom vcpkg registry, let us say `arrow` and `bloom-filter`, where `bloom-filter` depends on `arrow`
3. Create a project to consume the two packages in the vcpkg registry
    1. create`vcpkg.json` and `vcpkg-configuration.json`
    2. In `vcpkg-configuration.json`, it defines two registries, one for the built-in registry, and the other for the above custom registry on file system
    3. In `vcpkg.json`, it defines project `cycle-detected` depends on both `arrow` and `bloom-filter`
4. When installing dependencies for this `cycle-detected` project using `vcpkg install`, `cycle detected` error will be reported 
```
~/projects/cycle_detected
❯ vcpkg install
Error: Cycle detected during arrow:x64-osx:
bloom-filter:x64-osx
```
**Expected behavior**
Since `arrow` doesn't depend on `bloom-filter`, and `bloom-filter` depends on `arrow`, and the project `cycle-detected` depends on both of them, I don't see any reason why there is circular dependencies that cannot be resolved.

```
      cycle-detected-project ----->bloom-filter
                |_____________________|
                             ↓ 
                             arrow
```

**Failure logs** 
```
Error: Cycle detected during arrow:x64-osx:
bloom-filter:x64-osx
```
That is all the logs I can collect, let me know how I can collect more info for troubleshooting.

**Additional context**
The `arrow` and `bloom-filter` is just sample dependencies I used for reproducing this issue, in my real project, there are multiple such usages, aka. the project itself depends on two dependencies (pkg A/B), one of which (pkg A) depends on the other (pkg B). I would like to know if it is my incorrect usage or it is a bug of vcpkg. To my current understanding, this seems to be a valid usage.

**Comment by frivard-coveo** [comment]
We have observed the same issue internally. Has this bug been prioritized for a fix soon?

**Comment by Osyotr** [comment]
For a workaround, use overlay-port feature.

**Comment by ras0219-msft** [comment]
Sorry for the long time getting to this issue -- this looks to be a serious vcpkg bug and we will prioritize fixing it.

Thank you for the repro and bringing this to our attention!

**Comment by ras0219-msft** [comment]
We've determined that the actual cause of the issue is a difference between the version listed in the versions database and the version in the actual files on disk. The version database (and override) have `"port-version": 4`, but the actual files in the `ports/arrow/` directory are port version 0. We were able to resolve the issue by correcting the `ports/arrow/vcpkg.json`.

That said, the error message you received is obviously terrible and we'll look into fixing that.
